app/logical/bulk_related_tag_query.rb

```diff
@@ -0,0 +1,51 @@
+class BulkRelatedTagQuery
+  include ActiveModel::Serializers::JSON
+  include ActiveModel::Serializers::Xml
+
+  attr_reader :query, :category, :user
+
+  def initialize(query: nil, category: nil, user: nil)
+    @user = user
+    @query = Tag.normalize_query(query).split(' ').slice(0, 25)
+    @category = category
+  end
+
+  def tags
+    if category.present?
+      related_tags_by_category
+    elsif query.present?
+      related_tags
+    else
+      {}
+    end
+  end
+
+  def serializable_hash(**options)
+    tags
+  end
+
+  protected
+
+  def related_tags

```

Comment:
If I'm understanding this correctly, how about something like
```rb
def related_tags
  @related_tags ||= Tag.where(name: @query).each_with_object do |tag, hash|
    related    = tag.related_tag_array
    categories = Tag.categories_for(related.map(&:first))

    hash[tag.name] = related.map { |name, id| [name, id, categories.fetch(name, -1)] }
  end
end
```

Reply:
Oh, neat. Didn't know about each_with_object, that solves that nicely.